src/core/resource.cc

```diff
@@ -259,6 +264,33 @@ io_queue_topology::io_queue_topology(io_queue_topology&& o)
 
 }
 
+std::optional<std::vector<unsigned>>
+resource::internal::cpus_for_numa_node_id(hwloc::internal::topology_holder& topo_holder, unsigned numa_node_id) {
```

Comment:
The resource::allocate() already calculates the numa topology, it's worth returning it in the `resource::resources`

Reply:
I think I considered it, but forgot the details. Will look again.

Reply:
Ok, I considered something slightly different - to return the actual NUMA layout. But this can't be done until memory::configure() is run and can associate addresses.

I could return cpus_to_numa_node_id in resources, but I don't think that's its role. hwloc is responsible for answering random questions about topology, and since we want to abstract it, we put the helpers in resources.cc. Should resource::resources be enlarged each time we have a new query for topology?

Reply:
> Ok, I considered something slightly different - to return the actual NUMA layout. But this can't be done until memory::configure() is run and can associate addresses.

But you don't need addresses association here, only cpu:numa_node map

> I could return cpus_to_numa_node_id in resources, but I don't think that's its role. ... Should resource::resources be enlarged each time we have a new query for topology?

There are two options here. First -- make resources.cc be a bunch of transparent wrappers over hwloc so that we could ask it anything about topology, it would consult with hwloc behind the scenes, then return back some internal (btw, resources is not in internal) data type. Second -- make resources::allocate() talk to hwloc and return back anything that reactor would ever need to start. So far the second approach is in use.

Reply:
I refreshed my memory. I don't think this fits resources::allocate(). resources::allocate() says "look at system resources and tell me how they should be allocated to shards", not "look at system resources and tell me how they should be allocated to shards plus some random information".

Reply:
Then io-topology allocation deserves being removed from this method either (definitely not in this PR, unless you feel enough courage)

Reply:
I guess it's better to avoid churn, and define allocate() as "look at system resources and tell me how they should be allocated to shards, and give me any supporting topology information that can help me later".

In the end it's an internal API, even though we forgot to put it in the internal namespace.